tls/s2n_fingerprint.c

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_fingerprint.h"
+
+#include "crypto/s2n_fips.h"
+#include "crypto/s2n_hash.h"
+#include "stuffer/s2n_stuffer.h"
+#include "tls/extensions/s2n_extension_list.h"
+#include "tls/s2n_client_hello.h"
+#include "tls/s2n_crypto_constants.h"
+#include "utils/s2n_blob.h"
+#include "utils/s2n_result.h"
+#include "utils/s2n_safety.h"
+
+#define S2N_JA3_FIELD_DIV ','
+#define S2N_JA3_LIST_DIV  '-'
+
+/* UINT16_MAX == 65535 */
+#define S2N_UINT16_STR_MAX_SIZE 5
+
+/* clang-format off */
+const bool s2n_grease_values[] = {
+        [0x0A] = true, [0x1A] = true, [0x2A] = true, [0x3A] = true,
+        [0x4A] = true, [0x5A] = true, [0x6A] = true, [0x7A] = true,
+        [0x8A] = true, [0x9A] = true, [0xAA] = true, [0xBA] = true,
+        [0xCA] = true, [0xDA] = true, [0xEA] = true, [0xFA] = true,
+};
+/* clang-format on */
+
+/* Both bytes of a GREASE IANA value will be identical, and will
+ * match one of the values in s2n_grease_values.
+ *
+ * See https://datatracker.ietf.org/doc/html/rfc8701
+ */
+static S2N_RESULT s2n_assert_grease_value(uint16_t val)
+{
+    uint8_t byte1 = val >> 8;
+    uint8_t byte2 = val & 0x00FF;
+    RESULT_ENSURE_EQ(byte1, byte2);
+    RESULT_ENSURE_LT(byte1, s2n_array_len(s2n_grease_values));
+    RESULT_ENSURE_EQ(s2n_grease_values[byte1], true);
+    return S2N_RESULT_OK;
+}
+
+static bool s2n_is_grease_value(uint16_t val)
+{
+    return s2n_result_is_ok(s2n_assert_grease_value(val));
+}
+
+static S2N_RESULT s2n_fingerprint_hash_flush(struct s2n_hash_state *hash, struct s2n_stuffer *in)
+{
+    if (hash == NULL) {
+        RESULT_BAIL(S2N_ERR_INSUFFICIENT_MEM_SIZE);
```

Comment:
I wanted a usage error, since the problem is that the caller didn't give us a larger enough output buffer. And the message associated with this one seemed exactly right to me: "The provided buffer size is not large enough to contain the output data. Try increasing the allocation size."

But it might be unclear that this occurs because the output buffer isn't large enough. I'll add a comment.